cibuildwheel/projectfile.py

```diff
@@ -0,0 +1,50 @@
+from configparser import ConfigParser
+from pathlib import Path
+from typing import Any, MutableMapping, Optional
+
+import toml
+from packaging.specifiers import SpecifierSet
+
+
+class NoProjectFileError(RuntimeError):
+    pass
+
+
+class ProjectFile:
```

Comment:
To me, this file feels a bit over-engineered. It makes sense in a huge project, but I've always liked that `cibuildwheel` is reasonably straightforward, more or less reads front-to-back what's happening, and doesn't have a dozen files spread around.

I currently don't really see the advantage of this whole class over just keeping the old check for the existence of the file, and then a simple function `utils.get_project_requires_python(package_dir)`. It's not like reading this file is going to cause `cibuildwheel` to become very slow/inefficient?

Reply:
(For the record: I'm just being skeptical/questioning, here; I'm of course open to be convinced (or outvoted ;-) ) of the advantages of this construct, but I'm currently not sure this extra complexity, with custom exceptions etc, is worth it.)

Reply:
Personally, I like having small files with clear names, usually one per main class (helper classes can share files, as can support functions). If I see a class called ProjectFile, I'd like to have a file called projectfile.py. That way, you have a clear, focused set of includes instead of everything jumbled, you don't have items too heavily mixed, etc. `utils.py` is getting a bit crowded. I'd like to see a `architecture.py` with Architecture in it too, I think.

For _logic_, you want it to flow in one place. For classes, you want them in separate files, since they encapsulate logic. As long as the class is unit tested (which I could add) and you can trust it to work, then it simplifies the flow of logic where it is used.

Github even crosslinks if you click the class/function name, so I don't think it's too bad. :)

The design was general so that if we wanted to add more (like a setup.py parser, reading pyproject.toml to get config values, etc), it doesn't cause utils to grow, and all "project" related code could live together - existence checks and reading the files. Note that this always reads the files if present, while the old design was lazy and didn't read if nothing was needed (could use this and make it lazy, too, of course).

Custom exceptions are good! They let you be more specific when catching, and are extremely simple.

Reply:
> If I see a class called ProjectFile, I'd like to have a file called projectfile.py.

That can make sense for a class this size, yes, but I'm doubting the overegineering of making this a class, vs. just the

```
if any_of_these_files_exist:
    exit_with_same_error_as_now()

# ... later, when reading options

requires_python = None if requires_python_str is None else SpecifierSet(requires_python_str)
if requires_python is None:
    requires_python = utils.get_requires_python(package_dir)
```

Especially the first part is a lot simpler than wrapping `project_file = ProjectFile(package_dir)` in a try-except and catching a new one-time-use exception. It's a simple check, so why not use `if`? I'm currently not convinced the extra complexity is worth it. It makes perfect sense in a large 50+ files project, but not for the current state of `cibuildwheel` (where we've been aiming to keep things transparent and procedural, I believe), to me.

Reply:
I think it makes sense to have a `.exists()` method that checks this, that would remove the exception and allow you to put the check wherever you wanted it to go.

Reply:
Let me try a redesign after the arch stuff goes in, and see if you like it better.

Reply:
> I don't think utils is intended to read top to bottom.

No, utils isn't, ofc.

Maybe @joerick disagrees with me, and says this is more future-roof, though? It seems a bit heavy to me, right now, but ... I don't know what's coming later.

Reply:
Let me rewrite, and see if you like it better, I have some ideas. It was @joerick's original request that this be factored out, I had everything inline, originally.

Reply:
Oh, I missed that. You mean https://github.com/joerick/cibuildwheel/pull/536#discussion_r561357274?

(Which does suggest "as a function", but OK, let's not split hairs too much :-P )

Reply:
"or something", and classes are better than functions :) unless it's just one shot; and this can at least combine existence check and config reading, and if we eventually add reading anything else from the config at all, it should be refactored to a class, so might as well do it here.

Reply:
I haven't read all of the arguments above, but I don't think this warrants a class either. The trouble with classes is that they carry around hidden state, so they'd better be a good abstraction, otherwise you need to dig in and understand it and by then it's a net negative because you've got the inner workings of the class to worry about, as well as the task at hand.

I also think that generally, it's better to keep state in the filesystem, rather than in objects - that's a reason we don't have a `Project` object in cibuildwheel - it's just gonna read it from the filesystem anyway, so why try encapsulate that?

This coding style does feel unusual for some people (even sloppy), but we're optimising for lots of first-time readers, and they'll mostly be reading our code to understand why their build is failing. Literal is good.

As for the factoring out, a function seems good to me, because it can have an obvious name, that encapsulates the actual mechanism `read_package_python_requires(path)`.  But admittedly there is a little subjectivity to that!